## Objective:

The objective of this task was to use crime data from Chicago's Data Portal (Crimes - 2001 to present) to train a supervised learning model that could predict the occurrence of a crime in Chicago.

In [2]:
from sklearn import linear_model
import pandas as pd
import random

# The data to load
f = "crime_data/Crimes_-_2001_to_present.csv"

# Code to load random sample of .csv
num_lines = sum(1 for l in open(f))
print(num_lines)
# Sample size - in this case ~20% - Anymore than this I run into memory issues
size = int(num_lines / 5)
skip_idx = random.sample(range(1, num_lines), num_lines - size)

# Read the data
## According to Chicago Data Portal, 'Community Areas' is current. I am going to remove 'Community Area'.
drops = ['Case Number', 'Block',  'Description',  'Location Description', 'Updated On', 'Community Area']
data = pd.read_csv(f, skiprows=skip_idx).drop(drops, axis=1)
data.columns = ['ID','Date','IUCR',
                     'Primary_Type','Arrest','Domestic', 'Beat', 'District',
                     'Ward', 'FBI_Code', 'X_Coordinate', 'Y_Coordinate',
                     'Year', 'Latitude', 'Longitude', 'Location', 'Historical_Wards',
                     'Zip Codes', 'Community_Areas', 'Census_Tracts', 'Wards', 'Boundaries_ZIP',
                     'Police_Dist', 'Police_Beats']
print(data['Primary_Type'].value_counts())



6856243
THEFT                                288892
BATTERY                              249885
CRIMINAL DAMAGE                      156209
NARCOTICS                            143466
OTHER OFFENSE                         85551
ASSAULT                               85414
BURGLARY                              78371
MOTOR VEHICLE THEFT                   63526
DECEPTIVE PRACTICE                    54219
ROBBERY                               51918
CRIMINAL TRESPASS                     39177
WEAPONS VIOLATION                     14575
PROSTITUTION                          13734
PUBLIC PEACE VIOLATION                 9795
OFFENSE INVOLVING CHILDREN             9351
CRIM SEXUAL ASSAULT                    5734
SEX OFFENSE                            5262
INTERFERENCE WITH PUBLIC OFFICER       3218
GAMBLING                               2930
LIQUOR LAW VIOLATION                   2745
ARSON                                  2189
HOMICIDE                               1935
KIDNAPPING              

The code above is part of the **eda.py** program. This program limits the crime data (*Crimes_-_2001_to_present.csv*) to just a random 20% sample of the original rows. Working with any larger of a sample resulted in memory issues. 
The next step was to get a sense of the types of crimes contained in these data. By looking at the "Primary Type" column I could see a standardized description of the type of crime each observation in the data represents. The data contain a wide array of offenses. To build a model that could predict every type of crime would require many different features. Thus, I felt it was appropriate to limit this exercise to just looking at one type of crime. I chose Burglary since it is a common offense in Chicago, and I felt I could develop an appropriate set of model features for predicting this type of crime.
These sample data were then limited to just burglaries and output to a csv file.

## Features to Predict Burglary:

The program **burglaries_predictor_variables.py** takes the dataset described in the last step and adds some columns that will be used as features for our model:

### 1.  Time Realted Features

* For each observation, how much time has elapsed since the last burglary in the same community area (in days)
* Dummy variable to indicate working hours of the day (8am - 7pm)
* Dummy variable to indicate colder (winter) months (October to March)
* Dummy variable to indicate work days of the week (M - F)

### 2. Geographical Features (Community Area Features)

* Average amount of time that elapses between bruglaries within a community area (in days)
* Dummy variables to indicate each community area
* Total number of burglaries in that community area
* Number of affordable housing units within that community area
    * The data on affordable housing units by community area were obtained from here: https://data.cityofchicago.org/Community-Economic-Development/Affordable-Housing-Units-by-Community-Area/yvj4-y3fb

### 3. Law Enforcement Features

* Number of police beats (from crimes dataset)
* Haversine  distance to nearest police station (in kilometers)
    * The latitude and longitude for police station locations in Chicago were obtained from: https://data.cityofchicago.org/Public-Safety/Police-Stations/z8bn-74gv
    * Haversine distances were calculated from each burglary to each police station. Then for each burglary, the minimum of the distances calculated were used as the distance to the nearest police station.

## Random Forest Model:

I wanted to build a model that could predict the occurrence of a burglary within a certain future timeframe. More specifically I wanted to identify whether a burglary will happen in a community area within the next two weeks. The outcome variable is a binary outcome, taking on 1 if a burglary happened within 14 days of the last burglary in the community area, or 0 otherwise. Since the outcome is categorical, I would need to use a classifier algorithm. I chose to use a random forest model for several reasons.

The first reason I chose to use a random forest model was that it is easy to optimize the feature selection of the model. My approach, as shown below, would be to first run the random forest classifier with all the features I described in the previous section, score the algorithm, identify the most important features, then repeat the same classification algorithm using only the most important features, with gini-importance above a certain threshold.

The second reason I chose to use a random forest is that it resolves the issues of using a "greedy algorithm", like with implementing a decision tree, as we are using many decision trees to prevent our results from being biased by one decision tree outcome. In addition to this we are not susceptible to overfitting issues.

### Model With All Features:

In [2]:
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix
from sklearn.feature_selection import SelectFromModel
import pandas as pd
import numpy as np

n = 90
#
data = pd.read_csv("training_data_burglary.csv")
Training_Target = np.where((data['day_delta'] < 14), 1, 0)
Training_Data = data.iloc[:,np.r_[32:115]]

data_training, data_test, target_training, target_test = train_test_split(Training_Data, Training_Target, test_size = 0.25, random_state=1)
random_forest_machine = RandomForestClassifier(n_estimators=n)
random_forest_machine.fit(data_training, target_training)
predictions = random_forest_machine.predict(data_test)
print(accuracy_score(target_test, predictions))

cm = confusion_matrix(target_test,predictions)
confusion_matrix = pd.DataFrame(
	cm,
	columns = ['Predict No', 'Predict Yes'],
	index = ['True No', 'True Yes']
)
print(confusion_matrix)

0.8850717012276901
          Predict No  Predict Yes
True No          370         1888
True Yes         340        16788


The first model scored ~0.88507. As we can see with the model with all features, our random forest algorithm correctly predicted the occurrence of a burglary 16,788 times and failed to predict a burglary only 340 times. However, when burglaries didn't happen it tended to make a type I error and predict a burglary would be present, when in fact a burglary did not occur within the 14 day timespan.

### Feature Importance:

In [3]:
print(dict(zip(Training_Data.columns, random_forest_machine.feature_importances_)))

{'working_hours': 0.04789174542687778, 'winter': 0.046186494053002236, 'work_day': 0.04764561015895365, '10.0_area': 0.000298251158043319, '11.0_area': 0.002089878614287866, '12.0_area': 0.0035477153801159824, '13.0_area': 0.0007358934919541963, '14.0_area': 0.0007192470711610245, '15.0_area': 0.0001940991544416193, '16.0_area': 0.0002586259411464016, '17.0_area': 0.0007405095419473555, '18.0_area': 0.00407823727323944, '19.0_area': 0.00010158212693600333, '2.0_area': 0.00886888583788409, '20.0_area': 0.0002762760456842934, '21.0_area': 0.000751670997166664, '22.0_area': 0.00047848924929175604, '23.0_area': 0.0002953507295420248, '24.0_area': 0.0003191672594633944, '25.0_area': 0.0027972920320526733, '26.0_area': 0.0036152504435573393, '27.0_area': 0.0014157420329772032, '28.0_area': 0.0006771306899016116, '29.0_area': 0.0002795063129694009, '3.0_area': 0.004465321100533504, '30.0_area': 0.00033184716633579905, '31.0_area': 0.001247265813476658, '32.0_area': 0.0004347372358669323, '33.

The above output shows that a lot of our features aren't contributing much to the decision-making process for our algorithm. At first glance only a small number of features such as average time between burglaries (*CA_AVG_DELTA*) have a relatively high importance. The steps below will describe what was done to limit to only important features of the model.

In [4]:
### Train Model to Select Most Important Features and Refine The Model
# Fine Best Features
select_features = SelectFromModel(random_forest_machine, threshold = 0.1)
select_features.fit(data_training, target_training)
x_refined_train = select_features.transform(data_training)
x_refined_test = select_features.transform(data_test)

# Print Top Features
important_features = select_features.get_support()
feature_name = Training_Data.columns[important_features]
print(feature_name.value_counts())


police          1
CA_COUNT        1
CA_AVG_DELTA    1
dtype: int64


The above code does the following:
1) SelectFromModel detects the features of our previous model that have a feature importance above 0.1
2) The training and test data are then transformed to only take the features of the required importance
3) Print a list of the top features
    * Police Beat Count
    * Community Burglary Count
    * Community Average Time Between Burglaries

Below we will re-run the random forest classifier with only these three features.    

### Refined Model:

In [6]:
#Train and Test New Model
refined_forest_machine = RandomForestClassifier(n_estimators=n)
refined_forest_machine.fit(x_refined_train, target_training)
refined_predictions = refined_forest_machine.predict(x_refined_test)
print(accuracy_score(target_test, refined_predictions))

# for some reason i could not run this part of the code until i imported the confusion matrix library again
from sklearn.metrics import confusion_matrix
cmr = confusion_matrix(target_test,refined_predictions)
confusion_matrix = pd.DataFrame(
	cmr,
	columns = ['Predict No', 'Predict Yes'],
	index = ['True No', 'True Yes'])

print(confusion_matrix)


0.889817393995667
          Predict No  Predict Yes
True No          362         1896
True Yes         240        16888


The accuracy score for the refined model improved marginally to ~0.88981. The confusion matrix however indicates the model became better at predicting burglaries when they actually happened, but would still overpredict a burglary happening when it did not.